File: java/plugins/src/PlugletStreamInfo.cpp
/* 
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code is mozilla.org code.
 *
 * The Initial Developer of the Original Code is Sun Microsystems,
 * Inc. Portions created by Sun are
 * Copyright (C) 1999 Sun Microsystems, Inc. All
 * Rights Reserved.
 *
 * Contributor(s): 
 */
#include "PlugletStreamInfo.h"
#include "iPlugletEngine.h"
#include "nsCOMPtr.h"

jclass    PlugletStreamInfo::clazz = nsnull;
jmethodID PlugletStreamInfo::initMID = nsnull;

void PlugletStreamInfo::Initialize(void) {
    nsCOMPtr<iPlugletEngine> plugletEngine;
    nsresult rv = iPlugletEngine::GetInstance(getter_AddRefs(plugletEngine));
    if (NS_FAILED(rv)) {
	return;
    }
    JNIEnv *env = nsnull;
    rv = plugletEngine->GetJNIEnv(&env);
    if (NS_FAILED(rv)) {
	return;
    }
    clazz = env->FindClass("org/mozilla/pluglet/mozilla/PlugletStreamInfoImpl");
    if (env->ExceptionOccurred()) {
	env->ExceptionDescribe();
        clazz = nsnull;
	return;
    }
    clazz = (jclass) env->NewGlobalRef(clazz);
    initMID =  env->GetMethodID(clazz,"<init>","(J)V");
    if (!initMID) {
	env->ExceptionDescribe();
        clazz = nsnull;
	return;
    }
}

void PlugletStreamInfo::Destroy(void) {
    //nb who gonna call it?
    nsCOMPtr<iPlugletEngine> plugletEngine;
    nsresult rv = iPlugletEngine::GetInstance(getter_AddRefs(plugletEngine));
    if (NS_FAILED(rv)) {
	return;
    }
    JNIEnv *env = nsnull;
    rv = plugletEngine->GetJNIEnv(&env);
    if (NS_FAILED(rv)) {
	return;
    }
    if (clazz) {
	env->DeleteGlobalRef(clazz);
    }
}

jobject PlugletStreamInfo::GetJObject(const nsIPluginStreamInfo *streamInfo) {
    jobject res = nsnull;
    nsCOMPtr<iPlugletEngine> plugletEngine;
    nsresult rv = iPlugletEngine::GetInstance(getter_AddRefs(plugletEngine));
    if (NS_FAILED(rv)) {
	return nsnull;
    }
    JNIEnv *env = nsnull;
    rv = plugletEngine->GetJNIEnv(&env);
    if (NS_FAILED(rv)) {
	return nsnull;
    }
    if(!clazz) {
	Initialize();
	if (! clazz) {
	    return nsnull;
	}
    }
    res = env->NewObject(clazz,initMID,(jlong)streamInfo);
    if (env->ExceptionOccurred()) {
	env->ExceptionDescribe();
	res = nsnull;
    }
    return res;
}

File: java/plugins/classes/org/mozilla/pluglet/mozilla/PlugletStreamInfo.java
/* 
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code is mozilla.org code.
 *
 * The Initial Developer of the Original Code is Sun Microsystems,
 * Inc. Portions created by Sun are
 * Copyright (C) 1999 Sun Microsystems, Inc. All
 * Rights Reserved.
 *
 * Contributor(s): 
 */
package org.mozilla.pluglet.mozilla;

/**
 * This interface returns various information about the stream such as the
 * MIME type and whether the stream is seekable. 
 */
public interface PlugletStreamInfo {
    /**
     * Returns the MIME type for a particular stream.<p>
     * @return As stated above, returns the MIME type.
     */
    public String getContentType();
    /**
     * Indicates if a stream is seekable; that is, if it is possible
     * to move to a particular point in the stream.<p>
     * @return Returns true if the stream is seekable.
     */
    public boolean isSeekable();
    /**
     * Returns the length of a stream in bytes.
     */
    public int getLength();
    /**
     * Returns the time the data in the URL was last modified,
     * measured in seconds since 12:00 midnight, GMT, January 1, 1970.
     * <p>
     * @return Returns an integer value for the time since last
     * modification, as described above.
     */
    public int getLastModified();
    /** 
     * Specifies the URL that was used to originally request the stream.
     * <p>
     * @return Returns a String for the URL as described above.
     */ 
    public String getURL();
    /**
     * Requests reading from the input stream.
     */
    public void requestRead(ByteRanges ranges );
}


